claat/serve.go

```diff
@@ -0,0 +1,306 @@
+	"paper-material":                "PolymerElements/paper-material#1.0.6",
+	"paper-menu-button":             "PolymerElements/paper-menu-button#1.5.2",
+	"paper-menu":                    "PolymerElements/paper-menu#1.2.2",
+	"paper-radio-button":            "PolymerElements/paper-radio-button#1.3.1",
+	"paper-radio-group":             "PolymerElements/paper-radio-group#1.2.1",
+	"paper-ripple":                  "PolymerElements/paper-ripple#1.0.9",
+	"paper-scroll-header-panel":     "PolymerElements/paper-scroll-header-panel#1.0.16",
+	"paper-styles":                  "PolymerElements/paper-styles#1.2.0",
+	"paper-tabs":                    "PolymerElements/paper-tabs#1.8.0",
+	"paper-toast":                   "PolymerElements/paper-toast#1.3.0",
+	"paper-toolbar":                 "PolymerElements/paper-toolbar#1.1.7",
+	"platinum-sw":                   "PolymerElements/platinum-sw#1.2.4",
+	"polymer":                       "Polymer/polymer#1.4.0",
+	"sinonjs":                       "PolymerElements/sinonjs#1.17.1",
+	"stacky":                        "PolymerElements/stacky#1.3.2",
+	"webcomponentsjs":               "WebComponents/webcomponentsjs#0.7.24",
+}
+
+// overrideComp maps a given spec, as described in fetchRepo doc comments to a
+// desired component name.
+var overrideComp = map[string]string{
+	"googlecodelabs/codelab-components": "google-codelab-elements",
+	"google/code-prettify":              "google-prettify",
+}
+
+// fetchRepo downloads a repo from github and unpacks it into basedir/dest,
+// calling itself recursively for every dependency found in the unpacked bower.json.
+// If the basedir/dest directory already exists, fetchRepo does nothing.
+//
+// The spec has the following format: "githubuser/repo#version".
+// Any non-alphanumeric characters are stripped from the version.
+// If version is missing, "master" is used instead.
+// The bower registry is not used for naming resolution.
+func fetchRepo(basedir, spec string) error {
+	var user, repo, path, comp, ver string
+	s := strings.Split(spec, "#")
+	path = s[0]
+	if len(s) > 1 {
+		ver = s[1]
+	}
+	ver = strings.Trim(ver, "^~ ")
+	if ver == "" {
+		ver = "master"
+	}
+	s = strings.Split(path, "/")
+	user = s[0]
+	if len(s) > 1 {
+		repo = s[1]
+	}
+	// Check exception map to see if we should override the component name.
+	comp = repo
+	if c, ok := overrideComp[path]; ok {
+		comp = c
+	}
+	// if repo already exists locally, return immediately, we're done.
+	if _, err := os.Stat(basedir + "/" + comp); err == nil {
+		return nil
+	}
+	zipFile := basedir + "/" + comp + ".zip"
+	url := "https://github.com/" + user + "/" + repo + "/archive/v" + ver + ".zip"
+	err := downloadFile(zipFile, url)
+	if err != nil {
+		return err
+	}
+	// If get fails, it will download a file containing only "404: Not Found".
+	// We check for that case by looking for an unusually small file.
+	var st os.FileInfo
+	if st, err = os.Stat(zipFile); err != nil {
+		return err
+	}
+	if st.Size() < 20 {
```

Comment:
I would remove this hack in favour of checking response code from github. It will be 404 if the file is not there, so you won't even need to store anything on disk and check its size.

Reply:
Fixed.